### PR TITLE
[FrameworkBundle][Monolog] Added a new way to follow logs

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

----

If you want to try this PR, you can use [my fork](https://github.com/lyrixx/symfony-standard/tree/server-log):
```bash
git clone https://github.com/lyrixx/symfony-standard -b server-log symfony-se-logs
cd symfony-se-logs
composer install
bin/console server:start
bin/console server:log
```
and from anywhere `curl http://0:8000`

---

Basically, it's a new way to view and filter real time logs, from the CLI.

screenshot13

### CODE REVIEW

**Comment by yceruto** [comment]
:+1: Thanks you for this awesome feature! `--filter <expresion language>` :astonished: !!

Just a minor fix in your description to try this PR correctly, because [`server_log` handler is configured into `config_dev.yml`](https://github.com/lyrixx/symfony-standard/blob/server-log/app/config/config_dev.yml#L21) we need to run `curl http://0:8000/app_dev.php/` to see the output ;)

**Comment by lyrixx** [comment]
@yceruto By default, `bin/console` use the `dev` env. So the server is in `dev` env by default. That's why I did not add `/app_dev.php`, because it should not be necessary.

**Comment by yceruto** [comment]
>@yceruto By default, bin/console use the dev env. So the server is in dev env by default. That's why I did not add /app_dev.php, because it should not be necessary.

Right, but when you run `curl http://127.0.0.1:8000/` we are in prod env right? the handler is not there and the `server:log` doesn't shows nothing.

I missing something :/ ?

File: src/Symfony/Bundle/FrameworkBundle/Command/ServerLogCommand.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+        foreach ($logs as $message) {
+            $record = unserialize(base64_decode($message));
+
+            // Impossible to decode the message, give up.
+            if (false === $record) {
+                continue;
+            }
+
+            if ($filter && !$this->el->evaluate($filter, ['record' => $record])) {
```
Comment (maidmaid) [inline src/Symfony/Bundle/FrameworkBundle/Command/ServerLogCommand.php]: What about evaluate directly ``$record`` with ``->evaluate($filter, $record)``? ``filter`` option becomes easier to use: from ``--filter "record['level'] > 200"`` to ``--filter "level > 200"``

**Comment by yceruto** [comment]
> Right, but when you run curl http://127.0.0.1:8000/ we are in prod env right?

Nevermind, I forgot my env var in prod by default, sorry for the noise. This works as it is!

**Review by robfrawley — CHANGES_REQUESTED** [review]
You missed a bunch of short-array syntax instances. ;-) Thanks for this feature, though!

**Comment by hhamon** [comment]
Very nice feature!

**Comment by lyrixx** [comment]
@fabpot Should I move this command to the new server bundle?

File: src/Symfony/Bundle/FrameworkBundle/Command/ServerLogCommand.php [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            ->addOption('filter', null, InputOption::VALUE_REQUIRED, 'An expression to filter log. Example: record[\'level\'] > 200 or record[\'channel\'] in [\'app\', \'doctrine\']"')
+        ;
+    }
+
+    protected function initialize(InputInterface $input, OutputInterface $output)
+    {
+        $this->dumper = new CliDumper();
+        $this->dumper->setOutput($this->output = fopen('php://memory', 'r+b'));
+
+        $this->el = new ExpressionLanguage();
```
Comment (jakzal) [inline src/Symfony/Bundle/FrameworkBundle/Command/ServerLogCommand.php]: This property doesn't exist.

**Comment by fabpot** [comment]
@lyrixx Indeed, it should be moved to the new web server bundle.

**Comment by lyrixx** [comment]
I moved the command to the WebServerBundle.

**Comment by xabbuh** [comment]
The `ServerLogCommand` on one side and the `VarDumperFormatter` and `ServerLogHandler` on the other side are not wired. Is that intended? Wouldn't it be better to split this in two PRs then?

**Comment by lyrixx** [comment]
@xabbuh You are right, they are not wired. You might want to see my fork to see how it is wired.
IMHO, it's should not be split in 2 PRs because all theses classes work together.

**Comment by xabbuh** [comment]
@lyrixx Why did you close here?

**Comment by lyrixx** [comment]
Oups ... I don't know. Sorry for the noise.

**Comment by lyrixx** [comment]
@fabpot It looks like you are not totally convinced by this PR. I could add "experimental" (cf http://symfony.com/blog/experimental-features) if you want.

**Comment by nicolas-grekas** [comment]
I like this :) Here are some suggestions:
https://github.com/nicolas-grekas/symfony/commit/7b79e27b8545c2f34c0a7a3d87651196b223c57a

This patch fixes a 1s pause per log line I experienced when no log server was listening.
It uses persistent and asynchronous connections to make things faster.

**Comment by lyrixx** [comment]
> This patch fixes a 1s pause per log line I experienced when no log server was listening.

I'm unable to reproduce it. Do you have a reproducer ?

**Comment by lyrixx** [comment]
@nicolas-grekas I cherry-picked your commit and added a new one.

File: src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Symfony\Bridge\Monolog\Handler;
+
+use Monolog\Handler\AbstractHandler;
+use Monolog\Logger;
+use Symfony\Bridge\Monolog\Formatter\VarDumperFormatter;
+
+class ServerLogHandler extends AbstractHandler
```
Comment (yceruto) [inline src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php]: missing author?

File: src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Symfony\Bridge\Monolog\Handler;
```
Comment (yceruto) [inline src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php]: missing license

Comment (lyrixx) [inline src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php]: Thanks. Fixed (both comments)

File: src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            return;
+        }
+
+        stream_set_blocking($socket, false);
+
+        $keys = array('uid', 'uuid', 'log_uuid');
+        $logId = null;
+        foreach ($keys as $key) {
+            if (isset($record['extra'][$key])) {
+                $logId = $record['extra'][$key];
```
Comment (lyrixx) [inline src/Symfony/Bridge/Monolog/Handler/ServerLogHandler.php]: Fixed.

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    protected function configure()
+    {
+        $this
+            ->setName('server:log')
+            ->setDescription('Start a log server that displays logs in real time')
+            ->addOption('host', null, InputOption::VALUE_REQUIRED, 'The server host', '0:9911')
+            ->addOption('date-format', null, InputOption::VALUE_REQUIRED, 'The date format', 'H:i:s')
+            ->addOption('format', null, InputOption::VALUE_REQUIRED, 'The line format', '%s <%s>%-9s</> <options=bold>%-8.8s</> %s')
+            ->addOption('show-context', null, InputOption::VALUE_NONE, 'Display the context')
+            ->addOption('show-extra', null, InputOption::VALUE_NONE, 'Display the extra')
```
Comment (lyrixx) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: It does not use the expression language for that. I don't understand

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            ->addOption('filter', null, InputOption::VALUE_REQUIRED, 'An expression to filter log. Example: record[\'level\'] > 200 or record[\'channel\'] in [\'app\', \'doctrine\']"')
```
Comment (lyrixx) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: I guess we broke something. Thanks.

Comment (lyrixx) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: Actually, it's only `bin/console server:log --filter "level > 200"` ;) It's simpler.

Comment (stof) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: yeah, a review comment suggested to simplify it, but the description of the option was not updated when implementing the requested change

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    protected function configure()
+    {
+        $this
+            ->setName('server:log')
+            ->setDescription('Start a log server that displays logs in real time')
+            ->addOption('host', null, InputOption::VALUE_REQUIRED, 'The server host', '0:9911')
+            ->addOption('date-format', null, InputOption::VALUE_REQUIRED, 'The date format', 'H:i:s')
+            ->addOption('format', null, InputOption::VALUE_REQUIRED, 'The line format', '%s <%s>%-9s</> <options=bold>%-8.8s</> %s')
+            ->addOption('show-context', null, InputOption::VALUE_NONE, 'Display the context')
```
Comment (stof) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: should we really have options for that ? Couldn't it simply be based on the verbosity instead ? This seems like a perfect use case for it: display more details in verbose mode.

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            ->addOption('filter', null, InputOption::VALUE_REQUIRED, 'An expression to filter log. Example: record[\'level\'] > 200 or record[\'channel\'] in [\'app\', \'doctrine\']"')
+        ;
+    }
+
+    protected function initialize(InputInterface $input, OutputInterface $output)
+    {
+        $this->dumper = new CliDumper();
+        $this->dumper->setOutput($this->output = fopen('php://memory', 'r+b'));
+
+        if ($input->getOption('filter')) {
```
Comment (stof) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: this is wrong. ``initialize`` is called before the input validation is completed. so you should not start reading it here IMO

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        'white',
+        'yellow',
+    );
+
+    protected function configure()
+    {
+        $this
+            ->setName('server:log')
+            ->setDescription('Start a log server that displays logs in real time')
+            ->addOption('host', null, InputOption::VALUE_REQUIRED, 'The server host', '0:9911')
```
Comment (stof) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: why this default port ?

Comment (lyrixx) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: I really don't know.It was easy to type. What do you suggest ?

File: src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php (hunk 8 above)
Comment (lyrixx) [inline src/Symfony/Bundle/WebServerBundle/Command/ServerLogCommand.php]: Indeed I will remove it.

Actually I'm working on a much better default log display in command.

**Comment by fabpot** [comment]
@lyrixx any news?

**Comment by lyrixx** [comment]
@fabpot I want to merge #21705 first, then I will rebase this one.

**Comment by lyrixx** [comment]
Hello;

Here we go. I think this PR is ready for a last review.

I pushed everything to my fork for those who want to test it: https://github.com/lyrixx/symfony-standard/tree/server-log

File: src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php [hunk 11]
```diff
@@ -175,7 +178,9 @@ private function replacePlaceHolder(array $record)
 
         $replacements = array();
         foreach ($context as $k => $v) {
-            $replacements['{'.$k.'}'] = sprintf('<comment>%s</>', $this->dumpData($v, false));
+            // Remove quotes added by the dumper around string.
+            $v = trim($this->dumpData($v, false), '"');
+            $replacements['{'.$k.'}'] = sprintf('<comment>%s</>', $v);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php]: if the `$v` string ends with a backslash, will it break escaping?

**Review by nicolas-grekas — COMMENTED** [review]
I don't know how hard it'd be to functionally test that,
anyway, LGTM!

File: src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php (hunk 11 above)
Comment (lyrixx) [inline src/Symfony/Bridge/Monolog/Formatter/ConsoleFormatter.php]: Good catch.
I have fixed this issue by escaping the value.
Thanks.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by lyrixx** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @lyrixx.
